Mathlib/NumberTheory/SmoothNumbers.lean

```diff
@@ -0,0 +1,155 @@
+    exact ⟨fun H p hp ↦ not_succ_le_zero p (H p hp), fun H p hp ↦ False.elim <| H p hp⟩
+  ext m
+  rw [smoothNumbers, Set.mem_setOf, h, factors_eq_nil,
+      (show m ∈ ({1} : Set ℕ) ↔ m = 1 from Set.mem_def), Ne.def]
+  exact ⟨fun ⟨H₁, H₂⟩ ↦ H₂.resolve_left H₁, fun H ↦ ⟨H.symm ▸ one_ne_zero, Or.inr H⟩⟩
+
+/-- The product of the prime factors of `n` that are less than `N` is an `N`-smooth number. -/
+lemma prod_mem_smoothNumbers (n N : ℕ) : (n.factors.filter (· < N)).prod ∈ smoothNumbers N := by
+  have h₀ : (n.factors.filter (· < N)).prod ≠ 0
+  · simp only [ne_eq, List.prod_eq_zero_iff]
+    intro h
+    have H (p : ℕ) (h : p ∈ n.factors) : p ≠ 0 := Prime.ne_zero <| prime_of_mem_factors h
+    exact H 0 (List.mem_of_mem_filter h) rfl
+  refine ⟨h₀, fun p hp ↦ ?_⟩
+  have H := (mem_factors h₀).mp hp
+  obtain ⟨q, hq₁, hq₂⟩ := (Prime.dvd_prod_iff <| prime_iff.mp H.1).mp H.2
+  have hpq : p = q :=
+    (prime_dvd_prime_iff_eq H.1 <| prime_of_mem_factors <| List.mem_of_mem_filter hq₁).mp hq₂
+  refine hpq.symm ▸ ?_
+  simpa only [decide_eq_true_eq] using List.of_mem_filter hq₁
+
+/-- The sets of `N`-smooth and of `(N+1)`-smooth numbers are the same when `N` is not prime. -/
+lemma smoothNumbers_succ {N : ℕ} (hN : ¬ N.Prime) : N.succ.smoothNumbers = N.smoothNumbers := by
+  ext m
+  refine ⟨fun hm ↦ ⟨hm.1, fun p hp ↦ ?_⟩,
+         fun hm ↦ ⟨hm.1, fun p hp ↦ (hm.2 p hp).trans <| lt.base N⟩⟩
+  have H : p ≠ N := fun h ↦ hN <| h ▸ prime_of_mem_factors hp
+  exact Nat.lt_of_le_of_ne (lt_succ.mp <| hm.2 p hp) H
+
+/-- The non-zero noon-`N`-smooth numbers are `≥ N`. -/
+lemma smoothNumbers_compl (N : ℕ) : (N.smoothNumbers)ᶜ \ {0} ⊆ {n | N ≤ n} := by
+  intro n hn
+  have H := Set.mem_diff_singleton.mp hn
+  have H₁ := (Set.mem_compl_iff _ _).mp H.1
+  simp only [Nat.smoothNumbers, Set.mem_setOf_eq, not_and, not_forall, not_lt, exists_prop] at H₁
+  obtain ⟨m, hm₁, hm₂⟩ := H₁ H.2
+  exact hm₂.trans <| le_of_mem_factors hm₁
+
+/-- If `p` is a prime and `n` is `p`-smooth, then every product `p^e * n` is `(p+1)`-smooth. -/
+lemma Prime.pow_mul_mem_smoothNumbers {p n : ℕ} (hp : p.Prime) (e : ℕ) (hn : n ∈ smoothNumbers p) :
+    p ^ e * n ∈ smoothNumbers (succ p) := by
+  have hp' (e : ℕ) : p ^ e ≠ 0 := pow_ne_zero e hp.ne_zero
+  refine ⟨mul_ne_zero (hp' e) hn.1, fun q hq ↦ ?_⟩
+  rcases (mem_factors_mul (hp' e) hn.1).mp hq with H | H
+  · rw [mem_factors <| hp' e] at H
+    exact lt_succ.mpr <| le_of_dvd hp.pos <| Prime.dvd_of_dvd_pow H.1 H.2
+  · exact (hn.2 q H).trans <| lt.base p
+
+/-- If `p` is a prime and `n` is `p`-smooth, then `p` and `n` are coprime. -/
+lemma Prime.smoothNumbers_coprime {p n : ℕ} (hp : p.Prime) (hn : n ∈ smoothNumbers p) :
+    Nat.Coprime p n := by
+  rw [Prime.coprime_iff_not_dvd hp, ← mem_factors_iff_dvd hn.1 hp]
+  exact fun H ↦ lt_irrefl p <| hn.2 p H
+
+open List in
+/-- We establish the bijection from `ℕ × smoothNumbers p` to `smoothNumbers (p+1)`
+given by `(e, n) ↦ p^e * n` when `p` is a prime. -/
```

Comment:
It might be nice here to mention `smoothNumbers_succ` to show what happens if `p` isn't prime, and vice versa?

Reply:
OK.